[quant] Support 4-bit embedding_bag operators using the dtype quint4x2

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45881 [quant] Refactor qembeddingbag to remove duplicate code
* #45865 [quant] Support for 4-bit quantized EmbeddingBag module
* **#45752 [quant] Support 4-bit embedding_bag operators using the dtype quint4x2**
* #45751 [quant] Add 4-bit embedding_bag prepack/unpack support using quint4x2

Summary:
Use the torch.quint4x2 dtype to create 4-bit packed tensors in the previous PR.
These packed tensors can be directly consumed by the operator.
Serialization of the packed tensors is supported using torchbind custom class.
Module support will follow in a later PR.

Test Plan:
python test/test_quantization.py TestEmbeddingBagOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120996](https://our.internmc.facebook.com/intern/diff/D24120996)

supriyar committed:
[quant] Support 4-bit embedding_bag operators using the dtype quint4x2
Summary:
Use the torch.quint4x2 dtype to create 4-bit packed tensors in the previous PR.
These packed tensors can be directly consumed by the operator.
Serialization of the packed tensors is supported using torchbind custom class.
Module support will follow in a later PR.

Test Plan:
python test/test_quantization.py TestEmbeddingBagOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag.cpp:
```diff
@@ -11,6 +11,197 @@
+  auto offsets_data = offsets_new.data_ptr<int>();
+  const auto weight_data = weight.data_ptr<uint8_t>();
+  uint8_t* input_data = nullptr;
+  if (!weight.is_contiguous()) {
```
I think you can just do 
```
auto weight_contig = weight.contiguous();
input_data = weight_contig.data_ptr<uint8_t>();
```
without checking if weight is contigous or not here.

supriyar committed:
Update on "[quant] Support 4-bit embedding_bag operators using the dtype quint4x2"


Summary:
Use the torch.quint4x2 dtype to create 4-bit packed tensors in the previous PR.
These packed tensors can be directly consumed by the operator.
Serialization of the packed tensors is supported using torchbind custom class.
Module support will follow in a later PR.

Test Plan:
python test/test_quantization.py TestEmbeddingBagOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support 4-bit embedding_bag operators using the dtype quint4x2"


Summary:
Use the torch.quint4x2 dtype to create 4-bit packed tensors in the previous PR.
These packed tensors can be directly consumed by the operator.
Serialization of the packed tensors is supported using torchbind custom class.
Module support will follow in a later PR.

Test Plan:
python test/test_quantization.py TestEmbeddingBagOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120996](https://our.internmc.facebook.com/intern/diff/D24120996)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support 4-bit embedding_bag operators using the dtype quint4x2"


Summary:
Use the torch.quint4x2 dtype to create 4-bit packed tensors in the previous PR.
These packed tensors can be directly consumed by the operator.
Serialization of the packed tensors is supported using torchbind custom class.
Module support will follow in a later PR.

Test Plan:
python test/test_quantization.py TestEmbeddingBagOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120996](https://our.internmc.facebook.com/intern/diff/D24120996)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag.cpp:
```diff
@@ -11,6 +11,194 @@
+      offsets_in.has_value(),
+      "embedding_bag_4bit_rowwise_offsets expects offsets to be set");
+
+  TORCH_CHECK(weight.ndimension() == 2);
```
nit: I think ndimension is deprecated, dim() is the one we should be using

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag.cpp:
```diff
@@ -11,6 +11,194 @@
+
+  auto offsets_data = offsets_new.data_ptr<int>();
+  const auto weight_data = weight.data_ptr<uint8_t>();
+  uint8_t* input_data = nullptr;
```
nit: this line can be removed

jerryzh168 reviewed (approved):
looks good, left a few nit comments

supriyar committed:
Update on "[quant] Support 4-bit embedding_bag operators using the dtype quint4x2"


Summary:
Use the torch.quint4x2 dtype to create 4-bit packed tensors in the previous PR.
These packed tensors can be directly consumed by the operator.
Serialization of the packed tensors is supported using torchbind custom class.
Module support will follow in a later PR.

Test Plan:
python test/test_quantization.py TestEmbeddingBagOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120996](https://our.internmc.facebook.com/intern/diff/D24120996)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support 4-bit embedding_bag operators using the dtype quint4x2"


Summary:
Use the torch.quint4x2 dtype to create 4-bit packed tensors in the previous PR.
These packed tensors can be directly consumed by the operator.
Serialization of the packed tensors is supported using torchbind custom class.
Module support will follow in a later PR.

Test Plan:
python test/test_quantization.py TestEmbeddingBagOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24120996](https://our.internmc.facebook.com/intern/diff/D24120996)

[ghstack-poisoned]
